--- libevent-1.4.13-stable/buffer.c~	2009-09-25 00:20:20.000000000 +0200
+++ libevent-1.4.13-stable/buffer.c	2010-01-28 11:02:00.000000000 +0100
@@ -346,20 +346,21 @@
 
 #define EVBUFFER_MAX_READ	4096
 
 int
 evbuffer_read(struct evbuffer *buf, int fd, int howmuch)
 {
 	u_char *p;
 	size_t oldoff = buf->off;
 	int n = EVBUFFER_MAX_READ;
 
+        if (howmuch < 0 || buf->off + howmuch > buf->totallen) {  /**** pts ****/
 #if defined(FIONREAD)
 #ifdef WIN32
 	long lng = n;
 	if (ioctlsocket(fd, FIONREAD, &lng) == -1 || (n=lng) <= 0) {
 #else
 	if (ioctl(fd, FIONREAD, &n) == -1 || n <= 0) {
 #endif
 		n = EVBUFFER_MAX_READ;
 	} else if (n > EVBUFFER_MAX_READ && n > howmuch) {
 		/*
@@ -370,20 +371,21 @@
 		 * data we should read, we artifically limit it.
 		 */
 		if ((size_t)n > buf->totallen << 2)
 			n = buf->totallen << 2;
 		if (n < EVBUFFER_MAX_READ)
 			n = EVBUFFER_MAX_READ;
 	}
 #endif	
 	if (howmuch < 0 || howmuch > n)
 		howmuch = n;
+	}
 
 	/* If we don't have FIONREAD, we might waste some space here */
 	if (evbuffer_expand(buf, howmuch) == -1)
 		return (-1);
 
 	/* We can append new data at this point */
 	p = buf->buffer + buf->off;
 
 #ifndef WIN32
 	n = read(fd, p, howmuch);
